src/nlsolve/newton.jl

```diff
@@ -284,64 +358,103 @@ end
                 end
             end
         end
-        b = _vec(ztmp)
-    end
-
-    # update W
-    if W isa DiffEqBase.AbstractDiffEqLinearOperator
-        update_coefficients!(W, ustep, p, tstep)
-    end
-
-    if integrator.opts.adaptive
-        reltol = integrator.opts.reltol
-    else
-        reltol = eps(eltype(dz))
-    end
-
-    if is_always_new(nlsolver) || (iter == 1 && new_W)
-        linres = dolinsolve(integrator, linsolve; A = W, b = _vec(b), linu = _vec(dz),
-                            reltol = reltol)
-    else
-        linres = dolinsolve(integrator, linsolve; A = nothing, b = _vec(b), linu = _vec(dz),
-                            reltol = reltol)
+        _vec(ztmp)
     end
+    return b, ustep
+end
 
-    cache.linsolve = linres.cache
+## relax!
+relax!(dz::AbstractArray, nlsolver::AbstractNLSolver, integrator::DEIntegrator, f::TF) where {TF} = relax!(dz, nlsolver, integrator, f, relax(nlsolver))
+relax(dz::AbstractArray, nlsolver::AbstractNLSolver, integrator::DEIntegrator, f::TF) where {TF} = relax(dz, nlsolver, integrator, f, relax(nlsolver))
 
-    if DiffEqBase.has_destats(integrator)
-        integrator.destats.nsolve += 1
-    end
-
-    # relaxed Newton
-    # Diagonally Implicit Runge-Kutta Methods for Ordinary Differential
-    # Equations. A Review, by Christopher A. Kennedy and Mark H. Carpenter
-    # page 54.
-    if isdae
-        γdt = α * invγdt
-    else
-        γdt = γ * dt
-    end
-
-    !(W_γdt ≈ γdt) && (rmul!(dz, 2 / (1 + γdt / W_γdt)))
-    if (r = relax(nlsolver); !iszero(r))
+function relax!(dz::AbstractArray, nlsolver::AbstractNLSolver, integrator::DEIntegrator, f::TF, r::Number) where {TF}
+    if !iszero(r)
         rmul!(dz, 1 - r)
     end
+end
 
-    calculate_residuals!(atmp, dz, uprev, ustep, opts.abstol, opts.reltol,
-                         opts.internalnorm, t)
-    ndz = opts.internalnorm(atmp, t)
-    # NDF and BDF are special because the truncation error is directly
-    # proportional to the total displacement.
-    if integrator.alg isa QNDF
-        ndz *= error_constant(integrator, alg_order(integrator.alg))
+function relax!(dz::AbstractArray, nlsolver::AbstractNLSolver, integrator::DEIntegrator, f::TF, linesearch) where {TF}
```

Comment:
Have you verified type stability of this function?

Reply:
Fixed in [347d0e3](https://github.com/SciML/OrdinaryDiffEq.jl/pull/1792/commits/347d0e33aa3e7f1fb96872d942eb01fc69ad563b)